File: Sources/WippleLib/Parser/AST.swift
import Foundation

public class Box<T: Codable>: Codable {
    let value: T

    init(_ value: T) {
        self.value = value
    }

    required public init(from decoder: Decoder) throws {
        self.value = try T(from: decoder)
    }

    public func encode(to encoder: Encoder) throws {
        try self.value.encode(to: encoder)
    }
}

public struct AST: Codable {
    public let block: [[AST]]?
    public let list: [AST]?
    public let name: String?
    public let text: String?
    public let number: String?
    public let quoted: Box<AST>?
    public let location: Location
}

File: src/parser_bridge/ast.rs
use crate::parser_bridge::Location;
use serde::{Deserialize, Serialize};

#[derive(Clone, Serialize, Deserialize)]
pub struct AST {
    #[serde(flatten)]
    node: ASTNode,
    location: Location,
}

#[derive(Clone, Serialize, Deserialize)]
pub enum ASTNode {
    Block(Vec<Vec<AST>>),
    List(Vec<AST>),
    Name(String),
    Text(String),
    Number(String),
    Quoted(Box<AST>),
}
